File: 416.partition-equal-subset-sum.java
/*
 * @lc app=leetcode id=416 lang=java
 *
 * [416] Partition Equal Subset Sum
 */

// @lc code=start
class Solution {
    public boolean backtracking(int index, int[] nums, int total, int sum, int[][] cache) {
        if (total == sum) {
            return true;
        }
        if (index == nums.length || total > sum) {
            return false;
        }
        if (cache[index][total] != -1) {
            return cache[index][total] == 1 ? true : false;
        }
        cache[index][total] = (backtracking(index + 1, nums, total, sum, cache) || backtracking(index + 1, nums, total + nums[index], sum, cache)) == true ? 1 : 0;
        
        return cache[index][total] == 1 ? true : false;
    }

    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num: nums) {
            sum += num;
        }
        if (sum % 2 == 1) {
            return false;
        }
        int[][] cache = new int[nums.length][sum / 2 + 1];
        for (int i = 0; i < nums.length; i ++) {
            for (int j = 0; j <= sum / 2; j ++) {
                cache[i][j] = -1;
            }
        }
        return backtracking(0, nums, 0, sum / 2, cache);
    }
}
// @lc code=end

File: 416.partition-equal-subset-sum.py
#
# @lc app=leetcode id=416 lang=python3
#
# [416] Partition Equal Subset Sum
#

# @lc code=start
from typing import List

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total % 2: return False
        dp = {}
        def backtracking(index, current):
            if index == len(nums):
                return current == total // 2
            if (index, current) in dp:
                return dp[(index, current)]
            dp[(index, current)] = backtracking(index + 1, current) or backtracking(index + 1, current + nums[index])
            return dp[(index, current)]
        return backtracking(0, 0)
# @lc code=end

